foreach loops and sequence comprehensions (#9)
dafny-lang/rfcs: 0003-for-each.md

# Motivation
[motivation]: #motivation

Loops are notoriously difficult for Dafny users to verify, and coming up with the correct loop invariants to prove a loop correct 
is particularly challenging. The recent addition of `for` loops was a good step to abstract away from raw `while` loops,
but only supports consecutive integer indices. They still fall short of the expressiveness and economy of `for` or `foreach` loops in
many mainstream programming languages, which generally iterate through the contents of a collection of some kind.
It is a well-established best practice to avoid manual loop indexes wherever possible, as they force the assumption that datatypes
support efficient random access, and are more likely to be used incorrectly by accident:

```dafny
// Before:
method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
  for i := 0 to |s| {
    var left := s[i];
    for j := 0 to |s| {
      var right := s[i]; // Whoops!
      result := result + (left, right);  
    }
  }
}

// After:
method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {

>   foreach left | left in s {

Reviewer: Why not reuse the word "for"?

Author: As in still support the existing `for <variable> := <lo> to <hi> { ... }` syntax but also support `for <variables> | <range> { ... }`?

That should be doable in the parser AFAICT. I have a slight preference for how `foreach` reads but perhaps that's not worth the cost of yet another keyword. On the other hand communicating about the two different features will be easier if we can just say "for loops and foreach loops" versus "for loops over a range and for loops over a quantifier domain" or something like that. Open to opinions on this one.

Reviewer 2: I think it's better to use different keywords for `for` and `foreach` loops. Their loop invariants are different, and (depending on the enumerator used) the latter may have an effect on the state beyond just the index variable.

    foreach right | right in s {
      result := result + (left, right);
    }

>   }

Reviewer 3: It would also be good to have the following
```dafny
foreach left, right | left in s && right in s {
  result := result + (left, right);
}
```

Author: I had that thought too - that's equally valid (and does illustrate the power of the more general `foreach`) but I was a little wary of implying this is always the more idiomatic way to do it. I felt more confident it made sense in the sequence comprehension version.

}

// Or even better, as an expression:
function AllPairs(s: seq<nat>): seq<(nat, nat)> {

>   seq left, right | left in s && right in s

Reviewer: But this code doesn't compute the same thing as above, does it?  Is order guaranteed?

>   seq left, right | left in s && right in s

Reviewer 2: The proposed `seq` comprehension on L47 does not have an analogous `set` comprehension today, because a `set` comprehension with more than one bound variable is required to include a `:: Expr` at the end. To look more like what we have today, L47 would be better rendered as

```
seq left, right | left in s && right in s :: (left, right)
```

Author: Good catch, that's a typo. Will fix.

> }

Reviewer 3: Indeed, that would be a nice addition to the language.

If you want to make your example more compelling (meaning no comprehension can emulate your basic example, you could consider doing two operations at once, like computing an out variable "maximum".
```dafny
maximum := 0;
foreach left, right | left in s && right in s {
  result := result + (left, right);
  var tmpMax := left + right;
  maximum := if tmpMax > maximum then tmpMax else maximum;
}
```

Author: Good point, I'll put some thought into improving the motivating example - I think even the eventual user-facing documentation should have a good example of where a `foreach` loop is necessary because a sequence comprehension doesn't cut it.

Reviewer 4: Yeah, it seems to me that, even if `foreach` and sequence comprehensions were identical in what they could express, `foreach` would feel a lot more natural if the body is very imperative.

Author: This is part of why it feels right to me to design both features at once, as they are kind of two sides of the same fundamental coin: imperative and functional.

```

A more serious shortcoming is that there is currently no efficient way to iterate over the elements of a `set` at all.
The best alternative is to use the assign-such-that operator and set subtraction, illustrated in the reference manual as follows
([sec 10.5.2](https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#1052-sets)):

```dafny
// s is a set<int>
var ss := s;
while ss != {}

>   decreases |ss|

Reviewer 2: This `decreases` clause is not needed, since the default `decreases ss` works fine for this loop.

Author: Ha, don't tell me, tell the reference manual I copied it from. :) I can address it there separately.

{
  var i: int :| i in ss;
  ss := ss - {i};
  print i, "\n";
}
```

This is functionally correct, but because set subtraction requires creating a fresh copy,

> when compiled this loop requires quadratic time and memory (or at least garbage-collection work) in the size of the set.

Reviewer: Sounds like a bug in the runtime, no?

The only way to avoid this cost is to write additional external code in the target language to interface directly with the
internal runtime implementation of `set` values.

By comparison, the equivalent `foreach` loop is:

```dafny
foreach i | i in s {
  print i, "\n";

> }

Reviewer 3: That would be soo compelling. I hope you can break this down further and obtain an iterator for a set.

Author: It's a strong future possibility at least. Only supporting these higher-level operations over the elements of a set buys us a ton for now though, so I'm comfortable leaving it as future work.

Reviewer 4: I really want to be able to write those three lines, too!

I'd also love to see set iterators in the future, but agree that it's not necessary in the first iteration.

Reviewer: In line with what @samuelgruetter said above, though: since in Dafny `i in [1,2,3] <==> i in [3,2,1]`, shouldn't `foreach i | i in [1,2,3] { … }` be the same as `foreach i | i in [3,2,1] { … }`?

Reviewer 3: Since `foreach` is not an expression but a statement, its semantics require an ordering, and we would syntactically extract this ordering from the expression. In [my answer below](https://github.com/dafny-lang/rfcs/pull/9#issuecomment-1087668318), I explain how we could encode the ordering for the verifier, so that the two foreach statement you mention wouldn't be the same.

Reviewer: That's not right: `var s: seq<int> :| set(s) == {1,2,3}` is an expression, and it doesn't require an ordering.

Reviewer 3: It does not require ordering because there is only one statement to execute :-)
To be fair,  I should have said, _is a statement that will execute other statements in some orders_

```

The runtime implementation of this loop can use the native iteration features of the target language.

Going further, it is better to avoid writing any imperative loop at all where possible. Sticking to immutable data and

> expressions rather than statements allows logic to be used in specifications as well as implementation,

Reviewer: Does it really allow these things to be used in logic?  The problem with sets is that their iteration order is currently undefined, and so you cannot convert them to a sequence in the context of a `function method`.  You can do it in a function, but the result is not very easy to reason about because you don't know what iteration order it picked.

Reviewer 2: Up until this point in the document, there are several feedback questions about the ordering. This will be answered later in the document, I presume, but it would be good to add a sentence somewhere here in the motivation section that reassures the reader that order will be dealt with later, and for now the examples are more intended in a dream-like fashion ("wouldn't it be nice if you could write something like this--and don't think too hard about the ordering yet").

Author: I think that's a very good suggestion, and also that it would help to fill out more of the "sketch" of how that ordering is defined later on.

and reduces the burden on the Dafny verifier. The proposed sequence comprehension expression allows more
logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.

> Just producing a sequence of the values in a set is simple using a sequence comprehension expression: `seq i | i in s`.

Reviewer: Will that be allowed only in functions or also in function methods?

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## foreach Loops

A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
but is substantially more general and expressive. A very simple example only looks slightly different than expected:

```dafny

> foreach x | x in [1, 2, 3, 4, 5] {

Reviewer 5: What is the time complexity of computing the enumeration values? Suppose I have

```
foreach x | x in xs && x in ys {

}
```

What time complexity should I expect?

Author: Good question, and I should include that dimension in the more finalized definition of domain enumeration.

In that particular case it should be O(`|xs|` * `L`), where `L` is the cost of a `in` membership test in the `ys` collection. If `ys` is a set you'd expect it to be O(1) on average, but if it's a sequence you'd expect it to be O(`|ys|`).

In some cases we could likely optimize more cleverly, but we should be able to set upper bounds on the average and worst cases.

Reviewer 5: That `xs` serves as the value generating source here seems too implicit. You're saying that `x in ys && x in xs` could result in better performance if the membership test for `xs` is faster than for `ys`, but given that `&&` is usually a lazy operator, my intuition would be to put the fastest membership test first.

  print x, "\n";
}
```

The feature supports much more sophisticated uses as well, however,
including binding multiple variables at once and filtering:

```dafny
var myDoubleDeckerSeq: seq<seq<int>> := ...;

> foreach x, y | x in myDoubleDeckerSeq && y in x && y != 0 {

Reviewer 2: (I imagine that this particular range expression can be equivalently written as `y != 0 && y in x && x in myDoubleDeckerSeq`.)

Author: That version would be rejected, actually, since the rules for `&&` say that the enumeration is nested left-to-right. `y != 0` is enumerable (as a sub-domain of `int`) but not finite, and then `y in x` for each value of `y` would not even be enumerable (i.e. all possible sequences that contain `y`).

  Process(y);
}
```

The high-level syntax for a `foreach` loop reuses several existing concepts:

```
ForeachLoop ::=
  "foreach"
  QuantifierDomain
  LoopSpec

>   Body

Reviewer 2: I suggest also allowing `Body` to be omitted, as is allowed by `while` and `for` loops and by `forall` statements.

Author: I definitely have no objection, although I'm struggling to think of a compelling use case for that.

```

More specifically, this expands to:

```
ForeachLoop ::=
  "foreach" 
  Ident [":" Type] { "," Ident [":" Type] } [ "|" Expression ]
  { InvariantDecl | ModifiesDecl | DecreasesDecl }

>   Body

Reviewer 3: Would it be reasonable to consider that `while` loop is one of your `foreach` loops, under the following considerations:
- It's foreach loop with _no iteration variable_, not even one.
- The guard is the expression in the range
- Of course the _decreases clause_ is necessary.
?
If so, we might just want to replace the While loop internally by adding all these new iteration features.

Author: I don't think the semantics line up there correctly: if a loop with no bound variables was permitted, the semantics would be as I proposed further down, where the body would be executed once for every empty set of bindings produced in the domain's enumeration. That actually means such a `foreach` loop would be much closer to an `if` statement!

```dafny
var i := 0;
foreach | i <= 10 {
  // "i <= 10" is true, so the body would execute once
  print i, "\n";
  i := i + 1;
}
```

Your idea would only apply if you could quantify over the sequence of heap states somehow, I think.

You COULD translate `for` loops into `foreach` loops instead, but that will likely just make verifying the former cases more expensive, or at least add the burden of recognizing the special case of a domain where a single variable takes on those consecutive values. There was some discussion in the `for` loop RFC (#4) with the conclusion that there were benefits to having separate `for` and `foreach` loops, even if only to provide a simpler verification context for those learning Dafny.

```

A `foreach` loop closely resembles a `forall` statement, the key difference being that a `forall`
loop executes its body once for every tuple of quantified values simultaneously, whereas a `foreach` loop

> executes its body once for each tuple of quantified values in sequence, one at a time.

Reviewer 3: Note that `forall` only consist of these restrictions:
* A forall loop is always ghost (someone can confirm that? I have never seen a forall statement in a compiled context)
* A forall statement cannot contains statements that mutate variables.
* A forall statement cannot have breaks or continue.

```dafny
  ghost var result: seq<int> := []; // Would not 
  forall x | x in [1, 2, 3, 4, 5] {
   result := result + [x]; // Error
   continue; // Error
  }
```

Hence, with your proposal, we could simply make "forall" obsolete because foreach iterations that are in a ghost context, don't mutate variables, and without break or continue would have the SAME semantics as a forall statement.
Maybe the only difference would be the infiniteness. Forall would work for an infinite collection, but foreach wouldn't, right?

Author: No you can use `forall` in compiled contexts, but only in very limited circumstances. https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-forall-statement describes how there are three distinct cases of `forall` statements supported, and only the first can be used outside of ghost code.

I'm proposing `foreach` only allows a potentially infinite domain if not used in ghost code, and with either `decreases *` or an alternate termination measure. You make an interesting point though: the semantics of `forall` are that the body is executed simultaneously for every tuple, but if there is no interaction between these executions, that's indistinguishable from executing them sequentially.

For now I at least lean towards adding `foreach` while keeping `forall`, rather than allowing `forall` to be used in more cases: if nothing else I think `foreach` is a much better name for a loop where iterations CAN observe the effects of previous iterations. But I'll put some more thought into this.

Reviewer: The domain of a forall statement doesn't have to be enumerable, I think, so it's not superseded by `foreach` AFAICT.

Reviewer 2: Note that the `forall` statement is _not_ a loop. It is an _aggregate statement_. Its compiled form is like a generalization of simultaneous assignment. The beauty of the `forall` statement is that you use it _without a loop invariant_. :O This is made possible because--again--the `forall` statement is not a loop in the first place!

Loops (`while`, `for`, `foreach`) sequence their iterations. And to use them, you need a loop invariant, so that you can capture "what has been done so far".

Similarly to set comprehensions or assign-such-that statements, the domain of a `foreach` loop

> must be enumerable. The following loop would produce the error `"the domain of a foreach loop must be enumerable,

Reviewer 2: The domain has to be enumerable only if the set comprehension or assign-such-that statement is compiled. It is not a requirement in ghost contexts. Is that true also for `foreach`? It seems like a ghost `foreach` could support domains that are not enumerable, provided they still are ordered.

Author: We just chatted about this and came to the conclusion that a ghost `foreach` should still require an enumerable domain, since otherwise it's hard to imagine the semantics of a loop where the body can observe the effects of the executions before it.

Furthermore it should also be finite, since otherwise there isn't a well-defined notion of the state of the program "after" the infinite loop. This isn't the case for a `forall` statement since there is no looping, and the body executes once for each set of variable bindings simultaneously.

but Dafny's heuristics can't figure out how to produce an enumeration for 'x'"`.

```dafny

> foreach x: real | 0.0 <= x < 1.0 {

Reviewer: So this is not allowed, right?

Reviewer 2: I hope not.

Author: I'll add a comment to that effect just to be super clear :)

  ...
}
```

> The domain is allowed to be potentially infinite if the loop is used in a compiled context and an explicit `decreases` clause is provided.

Reviewer: What is the `domain` here?  If you have an invariant bounding the variable is that still an infinite domain?

Author: Good question. I think assuming that the invariant is allowed to refer to the bounds variables at all, I don't think it should be allowed to influence the domain enumeration.

`decreases *` is permitted, in which the `foreach` loop may never terminate. Any other `decreases` clause can be provided
to ensure the loop terminates even if the domain is potentially infinite. The following (slightly bizarre) example is legal:

```dafny
foreach x: nat 

>   decreases 10 - x

Reviewer 3: Shouldn't you add the invariant that x is less or equal to 10?

Author: Yup, this is what happens when you can't actually run the verifier to catch your mistakes. :) Thanks.

Reviewer 2: What should be added is not an invariant, but a range expression. There would be two problems with an invariant `x <= 10`. One problem would be that this does not hold true of all values that the `foreach` statement will consider (since this example used a `foreach` loop with no range expression). The other problem is that the loop invariant of a `foreach` statement is not allowed to mention the index variable (this is also true of the `for` statement).

Author: I think I need a better example here - it was questionable in the first place and trying to patch it with an invariant made it worse.

{
  print x, "\n";
  if x == 10 { break; }
}
```

Most `foreach` loops will take the form `foreach x | x in c { ... }`, where `c` is a `seq`, `set` or `multiset`.
This includes expressions like `m.Keys`, `m.Values` and `m.Items` when `m` is a `map`.
Looping over the keys and values of a map is particularly readable and aesthetically pleasing with this syntax,

> which also makes it easier for the compiled program to avoid unnecessary intermediate tuple values:

Reviewer: Isn't that going to be *less* efficient? The code below looks like it's doing one lookup per key, so you'll pay the lookup cost once per key in addition to the iteration cost, no?

Author: I originally had just `m[k] == v` before I realized I needed `k in m.Keys` as well, which not only hurts the elegance a bit but as you say also naturally implies a less efficient enumeration. I'm going to rework this - it should probably just be `k, v | (k, v) in m.Items` instead.

```dafny
foreach k, v | k in m.Keys && m[k] == v {
  print "m[", k, "] == ", v;
}

> ```

Reviewer 4: Oh, I like this!

Author: I liked it even more when I just had

```dafny
foreach k, v | m[k] == v {
  print "m[", k, "] == ", v;
}
```

but then realized I needed the `k in m.Keys` part. :)

Note that the range expression is optional, and if omitted the loop will iterate over all
possible values of the types of the bound variables. This is only permitted if all such types
are enumerable, which is not true of the `real` type, for example. This supports an elegant

> pattern for iterating over simple datatypes and other types with few values, including subset types.

Reviewer 3: That's a wonderful use case for testing !

Reviewer 4: Yeah, this is cool!

```
datatype Suit = Clubs | Diamonds | Hearts | Spades

> foreach s: Suit {

Reviewer 5: What happens if the type has an infinite size. How are the enumeration values generated?

Author: See my previous answer - does that cover it?

  ...
}

foreach b1: bool, b2: bool {
  expect TestMyFunctionWithOptions(b1, b2) == true;
}

type SmallIndex = x: nat | 0 <= x < 8

> foreach i: SmallIndex {

Reviewer 5: How are the enumeration values computed here? Is it possible to translate `x: nat | 0 <= x < 8` to something that efficiently creates the right values of `x` ? And what if the condition is more complex, does it become an enumeration over all natural numbers at some point?

Does the developer need to know the default order of basic types? Could there be scenarios where the developer has an incorrect assumption about the order?

Author: Is it definitely possible, and existing quantifiers and comprehensions already have a concept of "Bounded Pools" of values and how to iterate through them:

https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/AST/DafnyAst.cs#L11349

If you use a set comprehension like `set x: nat | 0 <= x < 8` in a compiled context, the resolver already detects that the range is a single `IntBoundedPool`, and the compiler will emit something like a `foreach` loop in the target language to build up the runtime set value: https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/Compilers/SinglePassCompiler.cs#L4671-L4686

With the existing features, if the heuristics can't figure out a finite pool to search through, you get an error. That will be the case here as well unless you're using a `foreach` loop with an explicit `decreases` clause.

You're right that for some cases users will want to know the enumeration ordering of some of these pools. I make the argument in other places that we should already document all of this better, since it affects the runtime performance of several existing Dafny features like the above. That will likely be a beneficial side-effect of this work.

  ...
}
```

There is no builtin support for automatically tracking the enumerated values so far, as there is for

> `iterator` values (see https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-iterator-types). 

Reviewer 2: This is going to be needed for every loop that you want to prove something about. It would be good to have built-in support for this.

Author: I am actually hoping that many such loops can be expressed as sequence comprehensions instead, where that isn't necessary. I'm open to using the same `xs` trick if we have to, but trying to leave it out of the initial scope.

It is straightforward to track these values manually, however:

```dafny
ghost var xs := [];
foreach x | x in s {
  ...
  xs := xs + [x];
}
```

It is even possible to attach this ghost state directly to the sequence with a helper function:

```dafny
foreach x, xs | (x, xs) in WithPrefixes(c) {
  ...
}
```

Where the signature of `WithPrefixes` is something like:

```dafny
function method WithPrefixes<T>(s: seq<T>): seq<(T, ghost seq<T>)> {
  ...
}
```

Similarly, a helper function can be provided to maintain a running index of the enumerated values:

```dafny
foreach x, i | (x, i) in WithIndexes(s) {
  ...
}

// Alternatively, if s is a sequence:

> foreach x, i | 0 <= i < |s| && s[i] == x {

Reviewer 3: NIce ! I was wondering how to iterate through a sequence with the index and the value from the beginning.

Author: Honestly I feel like this is more of an example of the power of using general quantifier domains to bind multiple variables, and less of an idiom I'd strongly encourage in general, since it's not much better than a `for` loop. :) It will be important for `WithIndexes` to be verification-friendly and runtime-efficient so we can move away from depending on random access.

Reviewer 3: It's actually still better than a for loop because you can do filtering on the value itself !

Author: True! I'm just not necessarily convinced that

```dafny
foreach x, i | 0 <= i < |s| && s[i] == x && x != 0 {
  ...
}
```

is significantly better than

```dafny
for i := 0 to |s| {
  var x := s[i];
  if x == 0 { continue; }
  ...
}
```

I suppose the one way it might be better is if you have a loop invariant that gets simpler when you can assume additional properties the domain gives you.

Reviewer 3: To me it's significantly better if you write it this way:

```
foreach i, x | 0 <= i < |s|
    && x == s[i]
    && x != 0 {
  ...
}
```

Indeed, everything that is an expression is almost always easier to reason about than a statement (for example, with the `for` and continue, you get the `x != 0` only implicitly after the block of the if because there is a continue, so you need to wrap your head to remember that x is not zero, whereas with the above, it's obvious)

  ...
}
```

## Sequence comprehensions

A sequence comprehension has identical syntax to a set comprehension, except that it begins with
`seq` rather than `set`. Therefore its syntax is:

```
SeqComprehension ::=
  "seq" 
  Ident [":" Type] { "," Ident [":" Type] } 
  "|" Expression
  [ :: Expression ]
```

Sequence comprehensions are the expression analogues of `foreach` loops, just as
`forall` expressions are the analogues of `forall` statements. The value of any given
`seq <BoundVariables> | <Range> :: <Term>` expression is specified by the following `foreach` loop:

```dafny
var result := [];
foreach <BoundVariables> | <Range> {
  result := result + [<Term>];
}
```

Sequence comprehensions are likely to be easier to work with than `foreach` loops, and Dafny users should be
encouraged to use them instead where possible. Many high-order operations on sequences can be expressed directly

> using sequence comprehensions:

Reviewer 3: Would it make sense to have two proposals, one for foreach loops, and one for sequence comprehensions? It would be good to figure out which one we should tackle first (meaning you :-) )

Author: I considered that too: I started trying to design `foreach` but then decided sequence comprehensions were a very closely related and arguably more appealing feature. Because their semantics are tightly coupled I'm inclined to discuss them together, even though in practice I plan to likely implement sequence comprehensions first as a separate contribution.

```dafny
// Filtering to optional values
var s: seq<Option<T>> := ...;
var filtered: seq<T> := seq x | x in s && x.Some? :: x.value;

// Zipping two lists together
var a: seq<T> := ...;
var b: seq<T> := ...;
assert |a| == |b|;
var zipped := seq i | 0 <= i < |a| :: (a[i], b[i])

// Map and then flatten (a.k.a. "FlatMap")
var f: T -> seq<T> := ...;
var c: seq<T> := ...;
var flatMapped: seq<T> := seq x, y | x in c && y in f(x) :: y;
```

> Note that the existing `seq(k, n => n + 1)` syntax is inconsistently-named in the Dafny reference manual, 

Reviewer 3: ```suggestion
Note that the existing `seq(size, i => i + 1)` syntax is inconsistently-named in the Dafny reference manual, 
```
Moreover, you can mention that this syntax to clone a sequence becomes complicated, because one has to add the precondition to the function:

`seq(|s|, i requires 0 <= i < |s| => s[i] + 1)`

but we refer to it here as a "sequence construction" expression, to disambiguate it from the proposed sequence comprehension feature.
Sequence comprehensions are strictly more expressive, as any construction `seq(size, f)` can be rewritten as 

> `seq i | 0 <= i < size :: f(i)`.

Reviewer 4: And we could eventually move `seq(size, f)` to a definition in the standard library, perhaps?

Author: Most likely - @RustanLeino was willing to fully deprecate the old version as long as we handled the equivalent sequence comprehension equally well in verification. Having an equivalent standard library version would make dropping support for it easier!

    * For sequences, the ordering is the obvious ordering of values in the sequence.
    * For (i)sets or multisets, the exact ordering is deterministic but unspecified.
    * Potentially-infinite sets are only allowed if they are enumerable.
    * If `x` is not a variable, then the enumeration produces an empty set of bindings once for each time the LHS appears in the RHS collection.
  * For an expression of the form `A && B`, the enumeration includes all bindings that satisfy both `A` and `B`. The enumeration ordering is defined by ordering first by `A`'s ordering and then by `B`'s, and the multiplicities of bindings is multiplied. See below for the pseudocode that calculates this enumeration.
    * This means that the `&&` operation is not fully symmetrical, but this is already true as verification considers it to be "short-circuiting": 
      `b != 0 && a/b > 1` is valid but `a/b > 1 && b != 0` is not.
    * This definition is chosen such that the multiplicities of results from `A && B` is the same as `B && A`, even if the ordering is not the same.

>   * For an expression of the form `A || B`, the enumeration will simply produce the values enumerated by `A` followed by those enumerated by `B`.

Reviewer 3: This case was not covered even before for set comprehensions, it would complain with "heuristics cannot figure out blablah".
The solution is to manually compute the sequence on which to iterate. For example, replacing `x in a || x in b` by `x in (a + b)`.
I think it would make sense to omit support for finding enumerations based on `||`, and the IDE could suggest such rewritings.

Author: True, the existing heuristics only consider conjuncts and give up on disjuncts. I may do as you say and trim disjuncts out of scope initially - I just wanted to make sure there was a reasonable answer to the semantics of enumerating disjuncts so we have a firm foundation. If and when there's need for it, though, I'd prefer to just implement it rather than have users rewrite their code.

Reviewer: @MikaelMayer But part of the motivation here is efficiency, and merging sets isn't a great idea for performance.

Author: Yup that's true. `a + b` at least strongly implies that a new value needs to be allocated at runtime, whereas enumerating `x in a || x in b` less so. There are definitely ways to optimize that (e.g. the lazy sequence concatenation logic in the C# and Java runtimes) but I like the idea of making it easier to infer the rough runtime cost just by looking at the structure of the code.

Reviewer 3: So if you wanted to enumerate the elements of `a + b` without duplication, you would explicitly write `x in a || (x in b && !x in a)`, correct?

Author: Yup that would work! It would only be efficient if both `a` and `b` had efficient `in` tests though.

  * If the context of an expression binds a variable that is not used in the expression, such as `seq x: bool | true`, 
    then the default ordering for the type of the variable is used (i.e. `[false, true]` in this case). 
    If this type is not enumerable, then the expression is not allowed.
  * For an expression of the form `!A`, the enumeration will produce all values of the types of the bound variables that do not satisfy `A`. 
    As above, if any of these types are not enumerable the expression is not allowed.

Here is the pseudocode for the semantics of the ordered enumeration of bindings for `A && B`:

```
IEnumerable<ActualBindings> EnumerateConjunction(Expr A, Expr B) {
  foreach (ActualBindings a in A) {
    Expr B' := B[a];  // Substitute the given bindings into the expression B
    foreach (ActualBindings b' in B') {
      yield return b';
    }
  }
}
```

And some illustrative examples using sequence comprehensions:

```dafny
// Special cases for one variable:

> seq x: T | A && B == seq x | A * seq x | B  // Not an actual built-in operation, but a generalization of set intersection.

Reviewer 3: ```suggestion
(seq x: T | A && B) == (seq x | A) * (seq x | B)  // Not an actual built-in operation, but a generalization of set intersection.
```

> seq x: T | A || B == seq x | A + seq x | B

Reviewer 3: ```suggestion
(seq x: T | A || B) == (seq x | A) + (seq x | B)
```

> seq x: T | !A == seq x: T - seq x | A       // Not an actual built-in operation, but a generalization of set subtraction.

Reviewer 3: ```suggestion
(seq x: T | !A) == (seq x: T) - (seq x | A)       // Not an actual built-in operation, but a generalization of set subtraction.
```

                                            // Only well-formed if T itself is enumerable.

> seq x | x in [1, 2, 2] && x in [2, 2, 1] == [1, 2, 2, 2, 2]

Reviewer 3: Why not 
```suggestion
seq x | x in [1, 2, 2] && x in [2, 2, 1] == [1, 2, 2, 2, 2, 1]
```
?
Also:
```suggestion
(seq x | x in [1, 2, 2] && x in [2, 2, 1]) == [1, 2, 2, 2, 2]
```
If you could fix parentheses below that'd be great. Parsing will always consider the seq expression to continue.

Author: Excellent question! The way I've defined the enumeration for conjuncts, the multiplicities are multiplied. The way this example breaks down is:

```dafny
seq x | x in [1, 2, 2] && x in [2, 2, 1] 
== seq x | x in [1] && x in [2, 2, 1] 
 + seq x | x in [2, 2] && x in [2, 2, 1]
== [1] 
 + [2, 2, 2, 2]
```

And thanks for catching all the missing parentheses. We'll see if I'm sufficiently annoyed by them to improve that. :)

Reviewer 3: Oh interesting. It's not obvious at all but it makes sense now. That would work the same way for a multiset, right?

and other existing features for loops in general,
as the semantics of such loops can usually be expressed as follows:

```dafny
// A loop of this form:
foreach x1, x2, ..., xN | <Range> {
  <Body>
}

// Is semantically equivalent to:
var __s := seq x1, x2, ..., xN | <Range> :: (x1, x2, ..., xN);
for __i := 0 to |__s| {
  var (x1, x2, ..., xN) := __s[__i];
  <Body>
}
```

This is not a complete equivalency as the domain of `foreach` loop with explicit `decreaes` clauses
are allowed to be infinite. It may make sense to support encoding potentially infinite sequences in the Boogie prelude
even if they are not (yet) supported in Dafny, to close this gap.

## Compilation

The most challenging part of compiling these new features is ensuring the enumeration
ordering of unordered collection types such as sets and multisets is deterministic when used in a sequence comprehension.
This is closely related to compiling let-such-that expressions (`var x :| ...; ...`), as such expressions
are implemented by enumerating through the parameters of a domain to search for a match,
and hence are influenced by this ordering. The current solution is to require any such expressions

> have only one solution, which is not an option here.

Reviewer: Why not?

As illustrated in [this paper](https://easychair.org/publications/paper/dM), the challenge is
that although the runtime values of `{1, 2}` and `{2, 1}` will compare equal to each other,
their internal representation may differ and hence their enumeration ordering may as well.

Most if not all current Dafny compilers implement a Dafny `set<T>` with some variation on a hash set,
which is generally a good default choice since this datatype offers constant-time access on average.

> Although different copies of the same set must place each value into the same hashing bucket, multiple

Reviewer: Why? Most hashset implementations are salted, and two copies of the same set put the same data in different buckets.

values may be assigned to the same bucket, so it is necessary to use a secondary data structure of some kind for each bucket.

> A common simple approach is to use a linked list,

Reviewer 5: > the requirement of a deterministic ordering based only on the set of contained elements.

I'm confused, I would think that either:
- You want to enumeration order over a collection to be defined, and then it must be deterministic as well
- You don't define the enumeration order, and then it doesn't have to be deterministic either since apparently you don't care about it, likely because the order doesn't affect the output of your program. You might be summing the elements of a set for example.

Author: You need the order to be deterministic so that, for example, the identity `S == S' ==> (seq x | x in S)[0] == (seq x | x in S')[0]` always holds. That's critical for verification to be sound and agree with compiled runtime behavior. But you don't actually have to know whether `seq x | x in {1, 2}` is actually `[1, 2]` or `[2, 1]`. That's what I mean by "under-specified".

Reviewer 5: Shouldn't the set to seq conversion `(seq x | x in S)` apply an implicit ordering on the seq, which for objects can use the pointer value as the key for that ordering?

Reviewer 5: To me the origin of the problem seems to be that you want to have an ordered iteration (like `(seq x | x in S)`) over an unordered collection.

Why not disallow an ordered iteration over an unordered collection?

in which case a single bucket will store elements in the order they were added and hence not meet the requirement
of a deterministic ordering based only on the set of contained elements.

One solution is to use a binary search tree as this secondary data structure instead, assuming that a total ordering of
the elements of the type `T` is available. Some hash set implementations, such as recent versions of the Java `HashSet<T>` type,
use a similar approach as an internal optimization when many hashing collisions occur, 

> so that the worst case runtime of lookups is O(N log N) rather than O(N).

Reviewer: Do you mean `O(log N)`?

Author: Yup :)

The included Dafny runtimes can provide an implementation of Dafny sets and multisets that use this approach consistently,

> and hence provide the efficient average-case access of a hash set but still provide a deterministic enumeration order.

Reviewer: I don't understand why this solves the problem, because iterating over the buckets still won't yield a consistent order.

The good news is that individual compilers are free to decide they do not support sequence comprehensions
whose ordering depends on any such collections. Dafny code that will be compiled using such compilers can 
fall back to using a `foreach` loop to iterate over sets instead, as statements are permitted to
observe non-determinism. This means a particular compiler can still map a Dafny `set<T>` value 
directly to a native implementation of sets, such as a Java `Set<T>`, even though these implementations will
enumerate their contents with non-deterministic ordering.

# Drawbacks
[drawbacks]: #drawbacks

The primary drawback to this proposal is the burden of adding the new concept of enumeration ordering. 
This means extra cognitive load for Dafny users, extra implementation effort for any new builtin collection types, 
and additional testing to ensure that all supported quantification domains are deterministically enumerable.

It can be argued that if any compiled Dafny program uses existing features that depend on domain enumeration,
such as `exists` expressions or `:|` statements, its runtime behavior already depends heavily on how this domain

> searching is implemented. The compiler only enforces that such domains are finite and enumerable, but an inefficient

Reviewer: That's true for `:|` with filters, but something like `var x :| x in s` isn't necessarily inefficient.

choice can lead to the search for a matching value taking orders of magnitude more time than expected.
Therefore we should already be documenting this aspect of Dafny semantics and implementation better anyway.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

The most obvious alternative is to only provide the simpler `foreach x in C` syntax, where `C` is any collection value.
This is certainly a well-formed and sound construct, but far less powerful than the proposed version, especially
when sequence comprehensions are excluded. It also still runs into some of the same semantic challenges around
ordering if `C` is allowed to be a set or multiset anyway, since verification of such a loop still needs to be aware of

> this ordering. Offering `foreach` loops along with sequence comprehensions means we can recommend the latter where possible,

Reviewer: I'm not sure: why isn't it enough to say that this may iterate in a different order every time?

with the former as a fallback that is inevitably more complicated to verify.

# Prior art
[prior-art]: #prior-art

Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 

> bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`

Reviewer 5: I find there's a big difference to Python's comprehensions and the ones introduced in this RFC, namely that in Python a value generator has to be specified, while in this RFC the comprehension seems to extract a value generator from a type and a boolean expression.

Author: Definitely. It's really only related prior art in that Python's comprehensions are values expressed in a similar way rather than loops. Sequence comprehensions are definitely more declarative and less explicit about the implementation.

Most mainstream programming languages include rich libraries for working with collections and iterators,
and define much the same operations such as "filter" and "flatMap".
The proposed Dafny features provide similar functionality at a higher level

> of abstraction more amenable to verification: quantification expressions are used to succinctly declare the behavior of

Reviewer: I don't know about more amenable to verification: Do we think that the verifier will be better at dealing with seq expressions and the like than at dealing with plain for loops and/or recursive functions?

Author: I should have said here that they should be friendlier for users, in terms of how much effort it takes to prove properties of a sequence comprehension compared to an equivalent loop or recursive function. My hope here is that the verifier will already know enough about an expression like `seq x | x in mySeq && P(x)` such that no additional work is necessary.

an operation that in spirit applies several such operations. See the "Future Possibilities" section for ideas for
pushing this philosophy even further.

The [JMatch language](https://www.cs.cornell.edu/andru/papers/padl03.pdf), an extension of Java to support
pattern matching, includes a `foreach` loop remarkably similar to the proposal here: it is parameterized by
a boolean expression and iterates through all matching bindings. For example, the statement `foreach(a[int i] != 0) n += i;`
sums the indices of all non-zero elements of the array `a`. The JMatch version declares the bound variables inline in
the boolean expression, which is also proposed here as a future possibility below.

> 

Reviewer 4: There's another instance I'm aware of that provides a capability not available in this proposal. I don't know if it's worth the complexity of supporting, but perhaps worth considering (or perhaps including in here to say why it's not the right thing to do).

[Cryptol](https://cryptol.net/) has two sequence comprehension separators, sequential and parallel. The sequential separators work like those in Python (and Haskell), and like `&&` in this proposal. For example, the cross product of two sequences is:

`crossProduct xs ys = [ (x, y) | x <- xs, y <- ys ]`

But you can also use a `|` instead of a comma to get this:

`zip xs ys = [ (x, y) | x <- xs | y <- ys ]`

You can also get this by zipping two iterators together first and then iterating over the result, which isn't much harder. The slightly more powerful (but sometimes confusing!) use case is when you alternate `,` and `|` in arbitrary ways.

Author: Interesting! Cryptol's version is at least worth naming as prior art as it's even closer to this proposal - thanks for the pointer!

My plan is more likely to handle that case when we support user-defined collection types. I included zipping as a key higher-order operation in [these experiments](https://github.com/dafny-lang/libraries/pull/37/files#diff-165e6a6e1f52e2d36e8c33448d2a99bc1553b21457565d3995247ed626a3e4faR490), and I'm fairly convinced it's maximally "Dafny-like" to require the LHS and RHS have provably-equal lengths. Between that and the fact that there's no natural boolean operator to express zipping, I'd say it's it doesn't fit as well here.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

One large open question is how difficult it will be to convince the verifier the following is correct,
since it will be a common idiom for depending on the unpredictable enumeration ordering of a set:

```dafny
function method SortedElements(s: set<int>): seq<int> 
  ensures multiset(SortedElements(s)) == multiset(s)
  ensures Seq.IsSorted(SortedElements(s), (a, b) => a < b)
{
  // Assume Seq.Sorted and Seq.IsSorted are part of the standard library
  Seq.Sorted(seq x | x in s, (a, b) => a < b)
}

method Foo() {
  var smallSet := {1, 2};
  var smallSortedSeq := SortedElements(smallSet);
  // smallSortedSeq is fully specified, but can the verifier figure that out?
  assert smallSortedSeq == [1, 2];
}
```

One of the first milestones when implementing this RFC will be to prototype just enough verification
plumbing to evaluate how difficult proving such properties will be, and how many additional axioms

> may need to be added to the Dafny prelude.

Reviewer: Here's a proof with Dafny as of today:

```dafny
predicate method IsLower(s: set<int>, m: int) {
  forall y :: y in s ==> m <= y
}

function Min(s: set<int>) : (m: int)
  requires s != {}
  ensures m in s
  ensures IsLower(s, m)
{
  var m :| m in s;
  if y :| y in s && y < m then
    var s' := s - {m};
    assert s == s' + {m};
    Min(s')
  else m
}

predicate Increasing(s: seq<int>) {
  forall i, j | 0 <= i < j < |s| :: s[i] <= s[j]
}

lemma IncreasingConcat(hd: int, tl: seq<int>)
  requires Increasing(tl)
  requires forall x | x in tl :: hd <= x
  ensures Increasing([hd] + tl)
{
  var s := [hd] + tl;
  forall i, j | 0 <= i < j < |s| ensures s[i] <= s[j] {
    assert s[j] in tl;
  }
}

lemma IsLowerSeq(s: set<int>, sq: seq<int>, m: int)
  requires IsLower(s, m)
  requires multiset(sq) == multiset(s)
  ensures forall x | x in sq :: m <= x
{
  forall x | x in sq ensures m <= x {
    assert x in multiset(sq);
  }
}

function method Sort(sq: seq<int>) : (sq': seq<int>)
  ensures Increasing(sq')
  ensures multiset(sq') == multiset(sq)

lemma MultiSetAddInj<T>(b: multiset<T>, a1: multiset<T>, a2: multiset<T>)
  requires b + a1 == b + a2
  ensures a1 == a2
{
  assert forall k :: a1[k] == a2[k] by {
    forall k {
      calc {
        a1[k];
        (b[k] + a1[k]) - b[k];
        (b + a1)[k] - b[k];
        (b + a2)[k] - b[k];
        (b[k] + a2[k]) - b[k];
        a2[k];
      }
    }
  }
}

lemma SortedElements_Unique(s: set<int>, sq: seq<int>)
  requires multiset(sq) == multiset(s)
  requires Increasing(sq)
  ensures sq == SortedElements(s)
{
  if sq == [] {
    assert |sq| == |multiset(s)| == 0;
  } else {
    var hd := sq[0];
    var sq' := sq[1..];
    var s' := s - {hd};
    calc { hd in sq; hd in multiset(sq); hd in multiset(s); hd in s; }

    assert sq == [hd] + sq';
    assert multiset(sq) == multiset([hd]) + multiset(sq');
    assert s == {hd} + s';
    assert multiset(sq) == multiset([hd]) + multiset(s');
    MultiSetAddInj(multiset([hd]), multiset(sq'), multiset(s'));

    assert Increasing(sq');
    SortedElements_Unique(s', sq');
  }
}

function SortedElements(s: set<int>): (sq: seq<int>)
  ensures multiset(sq) == multiset(s)
  ensures Increasing(sq)
{
  if s == {} then []
  else
    var m := Min(s);
    var s := s - {m};
    var sq := SortedElements(s);
    IsLowerSeq(s, sq, m); IncreasingConcat(m, sq);
    [m] + sq
} by method {
  sq := [];
  var leftover := s;
  ghost var removed := {};
  while leftover != {}
    decreases leftover
    invariant leftover !! removed
    invariant multiset(sq) == multiset(removed)
    invariant multiset(leftover + removed) == multiset(s)
  {
    var x :| x in leftover;
    leftover := leftover - {x};
    removed := removed + {x};
    sq := sq + [x];
  }
  sq := Sort(sq);
  SortedElements_Unique(s, sq);
}

method Foo() {
  var smallSet := {1, 2};
  var smallSortedSeq := SortedElements(smallSet);
  // smallSortedSeq is fully specified, but can the verifier figure that out?
  assert multiset(smallSortedSeq) == multiset{1, 2};
  assert smallSortedSeq == [1, 2];
}
```

Can we think about how that proof would change?

```

In general, it would allow omitting the list of bound variables in a quantifier domain expression and instead automatically
infer them from the list of free variables in the range expression.

## Unicode strings

As I have previously ranted about in https://github.com/dafny-lang/dafny/issues/413, the Dafny `string` type is currently an alias
for `seq<char>`, where `char` is defined as a UTF-16 code units, and hence allows invalid sequences according to the semantics of
Unicode. Modelling unicode correctly will involve specifying that the semantic contents of a UTF-8 string, for example, is
a sequence of Unicode code points that is not efficiently accessed with indices. The features proposed here provide ways to work with
such datatypes efficiently at runtime.

## More flexible conjunction enumerations

The proposed definition of domain ordering only allows the LHS of a conjunction to drive the primary enumeration,
but this could be loosened to support enumerating via the RHS and only ordering according to the LHS.
This would provide a succinct way to extract the values from a set in a fully predictable ordering, for example:

```dafny
var setOfReals := {3.141, 2.718, 1.618};

> var sortedList := seq x | x is real && x in setOfReals;

Reviewer: Why do you need `x is real` here?

Author: The idea was to provide a LHS that drives the desired ordering. I need to add this case (and more) to the ordering rules.

assert sortedList == {1.618, 2.718, 3.141};
```

## Array comprehensions

The same syntax for sequence comprehensions could also be used for flexible array initialization,
since this is another case where the user has to declare an ordered source of values:

```
var a := new int[10] i :: i * i;
var a := new int[|s|] i :: s[i];     // Where s is a seq<int>
var a := new int[] i | 10 <= i < 20;
```

It would likely be prudent to limit these cases to those where the size of the domain
is easy to infer statically, so that the size of the array to allocate is known before

> enumerating the values.

Reviewer: Or we could say that these are equivalent to sequence comprehension + cast to array?  I don't think we have a conversion to arrays from sequences, do we?

Author: Not directly, although it turns out there's a syntax I wasn't aware of similar to sequence constructions:

```dafny
var arr := new int[10] (i => i * i);
```

and not the opposite direction of aggregating multiple values into one.
The existing quantifier and comprehension expressions can be viewed as specific
cases of aggregating multiple values into a single result value.
An equally powerful feature would be to generalize this pattern and define
a parameterized collection expression to aggregate a quantifier domain into
a single value:

```dafny
collect(&&) x: T | P(x) :: Q(x)         == forall x: T | P(x) :: Q(x)
collect(||) x: T | P(x) :: Q(x)         == exists x: T | P(x) :: Q(x)
collect(SeqBuilder) x: T | P(x) :: Q(x) == seq x: T | P(x) :: Q(x)
collect(SetBuilder) x: T | P(x) :: Q(x) == set x: T | P(x) :: Q(x)
collect(+) x: T | P(x) :: Q(x)          == (summation)
collect(*) x: T | P(x) :: Q(x)          == (product)
collect(<) x: T | P(x) :: Q(x)          == (minimum)
collect(Averager) x: T | P(x) :: Q(x)   == (average)
collect(First(n)) x: T | P(x) :: Q(x)   == Take(seq x: T | P(x) :: Q(x), n)
...
```

This mirrors the `Collector` concept from the Java streaming APIs, and ideally the shape of the
parameter to `collect` expressions would define similar operations for merging intermediate results

> to leave the door open for parallel or even distributed computation.

Reviewer: Maybe we should look into the `loop` construct in lisp, which does all the above and more.  

https://lispcookbook.github.io/cl-cookbook/iteration.html

Author: Interesting, thanks for the link! That's roughly the shape of what I'm thinking of.
